contents/javascript/Javascript-framework-Redux.md

```diff
@@ -0,0 +1,226 @@
+  title: Redux
+  header: pagetop.md
+  footer: footer.md
+  head: head.md
+  siteNav: mainNav.md
+  pageNav: 3
+</frontmatter>
+
+<div class="website-content">
+  
+{{ booktitle | safe }}
+
+# Introduction to Redux
+
+**Author: [Labayna Neil Brian Narido](https://github.com/nbriannl)** <br>
+Reviewers: 
+
+## What is Redux?
+
+Simply put, Redux is library that manages application state for Javascript applications. You can use it together with view libraries such as React and Vue.
+
+More formally, the [official website](https://redux.js.org/introduction/getting-started) describes Redux as follows:
+
+> *Redux* is a predictable state container for JavaScript apps.
+>
+> It helps you write applications that behave consistently, run in different environments (client, server, and native), and are easy to test. On top of that, it provides a great developer experience, such as live code editing combined with a time traveling debugger.
+>
+> You can use Redux together with React, or with any other view library. It is tiny (2kB, including dependencies), but has a large ecosystem of addons available.
+
+### What is application state?
+
+([Adapted from this Medium article](https://medium.com/javascript-in-plain-english/the-only-introduction-to-redux-and-react-redux-youll-ever-need-8ce5da9e53c6))
+
+Think of it like a global object holding information that you will use for various purposes in your application. 
+
+For example:
+- For a general application you would need to keep store of whether a user is logged in and his/her user information
+- For a to-do list you would need to know what items are currently in the list
+- For a social media application you'd need to store serveral objects and arrays representing the stories to render in the News Feed
+
+All of the above can be considered as state.
+
+Modern web applications have large amount of state. And for complex apps, state is likely to be shared between components and updated from various parts of the code, making state less predictable. Which is why we need libraries like Redux to manage state. 
+
+## How does Redux work?
+
+### Store
+
+The whole application state is stored in a **single** store, a plain object. For example, the state of a todo app might look like this:
+
+```js
+{
+  todos: [{
+    text: 'Eat food',
+    completed: true
+  }, {
+    text: 'Exercise',
+    completed: false
+  }],
+  visibilityFilter: 'SHOW_COMPLETED'
+}
+```
+
+There are no direct setters to the store. The store is read-only; the only way to change the store is by emitting an action, which we will learn about next. 
+
+### Action
+
+Something in the state can be changed by dispatching an **action**, a plain object.
+
+```js
+{ type: 'ADD_TODO', text: 'Go to swimming pool' }
+{ type: 'TOGGLE_TODO', index: 1 }
+{ type: 'SET_VISIBILITY_FILTER', filter: 'SHOW_ALL' }
+```
+
+### Reducer
+
+The current state, and an action is passed into a **reducer**, a pure function which returns the next state of the application
```

Comment:
tooltip / popover / box on _pure function_ would be nice here, reader might not know what it is